enhancements/network/user-defined-network-segmentation.md

```diff
@@ -379,11 +382,264 @@ any user-defined primary network.
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Spec is immutable"
+    // +kubebuilder:validation:XValidation:rule=" self.topology != 'Layer3' || (self.topology == 'Layer3' && size(self.subnets) > 0)", message="Subnets is required for Layer3 topology"
+    // +kubebuilder:validation:XValidation:rule="self.topology != 'Localnet' || (self.topology == 'Localnet' && self.role == 'Secondary')", message="Topology localnet is not supported for primary network"
+    // +kubebuilder:validation:XValidation:rule="!has(self.ipamLifecycle) || (self.ipamLifecycle == 'Persistent' && (self.topology == 'Layer2' || self.topology == 'Localnet'))", message="ipamLifecycle is supported for Layer2 and Localnet topologies"
+    // +required
+    Spec UserDefinedNetworkSpec `json:"spec"`
+    // +optional
+    Status UserDefinedNetworkStatus `json:"status,omitempty"`
+}
+```
+Suggested API validation rules:
+- Spec defined as immutable
+  Avoid incomplete state in a scenario where a NAD is created according to UDN spec, and pods connected to the network, 
+  UDN spec changes, now pods connected to a network that was created from previous revision spec.
+- `Subents` are mandatory for `Layer3` topology.
+- `Localnet` topology is not supported for primary network.
+- `IPAMLifecycle` is supported for `Layer2` and `Localnet` topology.
+
+```golang
+// UserDefinedNetworkSpec defines the desired state of UserDefinedNetwork.
+type UserDefinedNetworkSpec struct {
+	// The topological configuration for the network.
+	// +kubebuilder:validation:Enum=Layer2;Layer3;Localnet
+	Topology NetworkTopology `json:"topology"`
+
+	// The network role in the pod (e.g.: Primary, Secondary).
+	// +kubebuilder:validation:Enum=Primary;Secondary
+	Role NetworkRole `json:"role"`
+
+	// The maximum transmission unit (MTU).
+	// The default value is 1400.
+	// +optional
+	Mtu uint `json:"mtu,omitempty"`
+
+	// The subnet to use for the network across the cluster.
+	// Only include the CIDR for the node. E.g. 10.100.200.0/24.
+	//
+	// IPv6 (2001:DBB::/64) and dual-stack (192.168.100.0/24,2001:DBB::/64) subnets are supported.
+	//
+	// When omitted, the logical switch implementing the network only provides layer 2 communication,
+	// and users must configure IP addresses for the pods.
+	// Port security only prevents MAC spoofing.
+	// +optional
+	Subnets []string `json:"subnets,omitempty"`
+
+	// A list of CIDRs.
+	// IP addresses are removed from the assignable IP address pool and are never passed to the pods.
+	// +optional
+	ExcludeSubnets []string `json:"excludeSubnets,omitempty"`
+
+    // Subnet used inside the OVN network topology.
+    // This field is ignored for non-primary networks (e.g.: Role Secondary).
+    // When omitted and network `role` is `primary` the following default CIDRs are used `100.65.0.0/24`, IPv6 `fd99::/64`.
```

Comment:
https://github.com/openshift/api/blame/ff31eaa0510fffdfbff1224b4091f1af3d3926b4/operator/v1/types_network.go#L549

Reply:
we could do something like
```
// Subnet used inside the OVN network topology.
// This field is ignored for non-primary networks (e.g.: Role Secondary).
// When omitted, this means no opinion and the platform is left to choose a reasonable default which is subject to change over time.
```
We do mention "current default value" here https://github.com/openshift/api/blame/ff31eaa0510fffdfbff1224b4091f1af3d3926b4/operator/v1/types_network.go#L549
but I don't think we can say for sure, since CRD version is not directly related to the ovn-k version, so CRD doesn't really know what is the "current" default as it depends on the ovn-k version.

Reply:
Done